File: python_proto/game.py
from time import sleep
import random
from players import Player, DumbPlayer, HumanPlayer, AIPlayer
from llm_client import GeminiClient, CerebrasClient
from collections import namedtuple
from table import Table, Card, Play
from typing import Optional, Type


class Round:
    def __init__(self, no: int, pass_direction: dict[str, str]) -> None:
        self.no = no
        self.current_trick: list[Play] = []
        self.trick_no = 0
        self.leads_trick: Optional[str] = None
        self.hearts_broken = False
        self.queen_of_spades_broken = True  # False if no < 1 else True
        self.pass_direction = pass_direction
        self.player_moonshot: Optional[str] = None


class Game:
    def __init__(self, player_ids: list[str], players: dict[str, Player]) -> None:
        self.player_ids = player_ids  # order matters
        self.players = players
        self.player_scores = {player_id: 0 for player_id in self.player_ids}

        self.table = Table(player_ids=[])
        self.round_no = -1

        # round data
        self.round = Round(-1, pass_direction={})
        self.old_rounds: list[Round] = []
        print(f"Starting game with {self.player_ids}")

    def _calculate_whose_move(self) -> str:
        assert isinstance(self.round.leads_trick, str)

        return self.player_ids[
            (
                self.player_ids.index(self.round.leads_trick)
                + len(self.round.current_trick)
            )
            % 4
        ]

    def _pass_cards(self, passed_by_player: dict[str, set[Card]]) -> None:
        updates: dict[str, dict[str, set[Card]]] = {
            player_id: {"added": set(), "removed": set()}
            for player_id in self.player_ids
        }
        for player_id, passed_cards in passed_by_player.items():
            assert not passed_cards - self.table.hands[player_id]
            updates[player_id]["removed"].update(passed_cards)
            updates[self.round.pass_direction[player_id]]["added"].update(passed_cards)

        self.table.update_hands(updates)

        for player_id, hand in self.table.hands.items():
            if Card("♣", "2") in hand:
                self.round.leads_trick = player_id

    def _start_round(self) -> None:
        self.round_no += 1

        pass_direction = {
            self.player_ids[i]: self.player_ids[(i + 1 + self.round_no % 3) % 4]
            for i in range(4)
        }
        if self.round:
            self.old_rounds.append(self.round)
        self.round = Round(no=self.round_no, pass_direction=pass_direction)

        print(
            f"Starting round {self.round_no}. Scores: {self.player_scores}. Pass directions: {pass_direction}."
        )

        self.table = Table(player_ids=self.player_ids)

        # deal 13 cards
        for player_id in self.player_ids:
            self.table.deal_cards(13, player_id)

    def _filter_illegal_options(self, hand: set[Card], player_id: str) -> set[Card]:
        hand = hand.copy()

        if Card("♣", "2") in hand:
            return {Card("♣", "2")}

        leading_suit = (
            self.round.current_trick[0][1].suit if self.round.current_trick else None
        )
        leading_suit_cards = {
            card for card in self.table.hands[player_id] if card.suit == leading_suit
        }
        if leading_suit and leading_suit_cards:
            return leading_suit_cards

        if self.round.trick_no == 0:
            hand = {
                card for card in hand if card.suit != "♥" and card != Card("♠", "Q")
            }
        if not self.round.current_trick and not self.round.hearts_broken:
            safe = {card for card in hand if card.suit != "♥"}
            if safe:
                hand = safe
        return hand

    def _finish_trick(self) -> None:
        leading_suit = self.round.current_trick[0].card.suit
        trick_taken_by = sorted(
            [
                (self.table.ranks.index(card.rank), player_id)
                for player_id, card in self.round.current_trick
                if card.suit == leading_suit
            ]
        )[-1][1]

        cards_in_trick = [card for player_id, card in self.round.current_trick]
        self.table.cards_taken[trick_taken_by] |= set(cards_in_trick)

        print(
            f"Trick {self.round.trick_no}: {self.round.current_trick} taken by {trick_taken_by}"
        )

        self.round.leads_trick = trick_taken_by
        self.round.current_trick = []
        self.round.trick_no += 1

    def _make_play(self, player_id: str, played_card: Card) -> None:
        assert player_id == self._calculate_whose_move()
        self.table.update_hands({player_id: {"removed": {played_card}}})
        self.round.current_trick.append(Play(player_id, played_card))

        if any([card.suit == "♥" for player_id, card in self.round.current_trick]):
            self.round.hearts_broken = True

    def _round_continues(self) -> bool:
        if any(self.table.hands.values()):
            return True
        return False

    def _check_moonshot(self) -> Optional[str]:
        for player_id in self.player_ids:
            if (
                sum(
                    [
                        1
                        for suit, rank in self.table.cards_taken[player_id]
                        if suit == "♥"
                    ]
                )
                == 13
                and Card("♠", "Q") in self.table.cards_taken[player_id]
            ):
                return player_id
        return None

    def _finish_round(self) -> None:
        player_moonshot = self._check_moonshot()
        if player_moonshot:
            self.round.player_moonshot = player_moonshot
            print(
                f"Player {player_moonshot} shot the moon! Adding 26 to everyone else's score."
            )
            for player_id in [pid for pid in self.player_ids if pid != player_moonshot]:
                self.player_scores[player_id] += 26
        else:
            for player_id, cards_taken in self.table.cards_taken.items():
                self.player_scores[player_id] += sum(
                    (self.table.card_values[card] for card in cards_taken)
                )
        print(f"Round {self.round_no} over, scores: {self.player_scores}")

    def _game_continues(self) -> bool:
        if (
            any([v > 100 for v in self.player_scores.values()])
            and list(self.player_scores.values()).count(
                min(self.player_scores.values())
            )
            == 1  # check if there's a single winner
        ):

            print(f"Game finished! Scores: {self.player_scores}")
            return False
        return True

    def play_round(self) -> None:
        self._start_round()

        passed_cards = {}
        for player_id, player in self.players.items():
            pass_recipient = self.round.pass_direction[player_id]
            passed_cards[player_id] = player.pass_three_cards(
                hand=self.table.hands[player_id], recipient=pass_recipient
            )
            print(f"{player_id} passed {passed_cards[player_id]} to {pass_recipient}")
        self._pass_cards(passed_cards)

        while self._round_continues():
            for player_id in self.player_ids:
                assert [isinstance(c, Card) for c in self.table.hands[player_id]]

            player_id = self._calculate_whose_move()
            player_order = [
                (self.player_ids.index(player_id) + i) % 4 for i in range(0, 4)
            ]
            print(
                f"Starting trick {self.round.trick_no}. Order: {[self.player_ids[i] for i in player_order]}"
            )
            assert (
                len(set(len(hand) for hand in self.table.hands.values())) == 1
            )  # assert everyone has the same number of cards in hand
            for _ in range(4):

                player_id = self._calculate_whose_move()
                player = self.players[player_id]

                print(f"{player_id} moves...")
                played_card = player.select_card_for_trick(
                    hand=self.table.hands[player_id],
                    legal_cards=self._filter_illegal_options(
                        self.table.hands[player_id], player_id
                    ),
                    trick=self.round.current_trick,
                    cards_taken=self.table.cards_taken,
                )
                self._make_play(player_id, played_card)

            self._finish_trick()

        self._finish_round()

    def play(self) -> None:
        while self._game_continues():
            self.play_round()


if __name__ == "__main__":
    llm_client = CerebrasClient()

    for i in range(10):
        player_classes: list[Type[Player]] = [AIPlayer, DumbPlayer, DumbPlayer, DumbPlayer]
        player_ids: list[str] = ["Tom", "Mary", "Holly", "Joslyn"]
        players: dict[str, Player] = {}
        for player_id, player_class in zip(player_ids, player_classes):
            players[player_id] = player_class(
                player_id=player_id,
                opponents=[pid for pid in player_ids if pid != player_id],
                llm_client=llm_client,
                count_cards=False,
                shoot_the_moon=False,
            )

        game = Game(player_ids=player_ids, players=players)
        game.play()

File: game.js
import { Table, RANKS } from "./table.js";
import {
  draw_trick_card,
  draw_hand,
  clear_table,
  draw_nametags,
  draw_status,
} from "./renderer.js";

class Round {
  constructor(no, pass_direction) {
    this.no = no;
    this.current_trick = [];
    this.trick_no = 0;
    this.leads_trick = null;
    this.hearts_broken = false;
    this.pass_direction = pass_direction;
    this.player_moonshot = null;
  }
}

export class Game {
  constructor(player_ids, players, renderer) {
    this.player_ids = player_ids;
    this.players = players;

    const sides = ["bottom", "left", "top", "right"];
    this.player_ids_to_sides = Object.fromEntries(
      sides.map((s, i) => [player_ids[i], s])
    );

    this.player_scores = new Map();
    player_ids.forEach((player_id) => this.player_scores.set(player_id, 0));

    this.table = new Table(player_ids);

    this.round_no = -1;
    this.round = null;
    this.old_rounds = [];
    this.renderer = renderer;
  }

  _calculate_whose_move() {
    return (
      (this.player_ids.indexOf(this.round.leads_trick) +
        this.round.current_trick.length) %
      4
    );
  }

  _pass_cards(passed_by_player) {
    const updates = new Map();
    this.player_ids.forEach((player_id) =>
      updates.set(
        player_id,
        new Map([
          ["added", []],
          ["removed", []],
        ])
      )
    );

    for (const [player_id, passed_cards] of passed_by_player.entries()) {
      updates.get(player_id).set("removed", passed_cards); // remove cards from the passer
      var recipient = this.round.pass_direction.get(player_id);
      updates.get(recipient).set("added", passed_cards); // add cards to the recipient
    }

    this.table.update_hands(updates);

    for (const [player_id, hand] of this.table.hands.entries()) {
      if (hand.includes("♣2")) {
        this.round.leads_trick = player_id;
      }
    }
  }

  _start_round() {
    this.round_no += 1;

    const pass_direction = new Map();
    for (var i = 0; i < 4; i++) {
      pass_direction.set(
        this.player_ids[i],
        this.player_ids[(i + 1 + (this.round_no % 3)) % 4]
      );
    }

    if (this.round !== null) {
      this.old_rounds.push(Round);
    }
    this.round = new Round(this.round_no, pass_direction);

    console.log(
      `Starting round ${this.round_no}. Scores: ${[...this.player_scores.entries()]}. Pass directions: ${[...pass_direction.entries()]}.`
    );

    this.table = new Table(this.player_ids);

    for (const player_id of this.player_ids) {
      this.table.deal_cards(13, player_id);
    }
  }

  _filter_illegal_options(hand, player_id) {
    hand = hand.slice();

    if (hand.includes("♣2")) {
      return ["♣2"];
    }

    const leading_suit =
      this.round.current_trick.length !== 0
        ? this.round.current_trick[0][1][0]
        : null;
    const leading_suit_cards = this.table.hands
      .get(player_id)
      .filter((card) => card[0] === leading_suit);

    if ((leading_suit !== null) & (leading_suit_cards.length !== 0)) {
      return leading_suit_cards;
    }

    if (this.round.trick_no === 0) {
      hand = hand.filter((card) => card[0] !== "♥");
      hand = hand.filter((card) => card !== "♠Q");
    }

    if ((this.round.current_trick.length === 0) & !this.round.hearts_broken) {
      const safe = hand.filter((card) => card[0] !== "♥");
      if (safe.length !== 0) {
        hand = safe;
      }
    }
    return hand;
  }

  _finish_trick() {
    const leading_suit = this.round.current_trick[0][1][0];

    const plays_followed_suit = this.round.current_trick.filter(
      (play) => play[1][0] === leading_suit
    );

    var highest_rank = -1;
    var trick_taken_by = null;

    for (const [player_id, card] of plays_followed_suit) {
      var current_rank = RANKS.indexOf(card[1]);
      if (highest_rank < current_rank) {
        highest_rank = current_rank;
        trick_taken_by = player_id;
      }
    }

    const cards_in_trick = this.round.current_trick.map((item) => item[1]);

    this.table.cards_taken.set(
      trick_taken_by,
      this.table.cards_taken.get(trick_taken_by).concat(cards_in_trick)
    );

    console.log(
      `Trick ${this.round.trick_no}: ${this.round.current_trick} taken by ${trick_taken_by}`
    );
    this.round.leads_trick = trick_taken_by;
    this.round.current_trick = [];
    this.round.trick_no += 1;
  }

  _make_play(player_id, played_card) {
    const updates = new Map([
      [
        player_id,
        new Map([
          ["added", []],
          ["removed", [played_card]],
        ]),
      ],
    ]);

    this.table.update_hands(updates);
    this.round.current_trick.push([player_id, played_card]);

    const hearts_played = this.round.current_trick.filter(
      ([player_id, played_card]) => played_card[0] === "♥"
    );
    if (hearts_played.length !== 0) {
      this.round.hearts_broken = true;
    }
  }

  _round_continues() {
    for (const hand of this.table.hands.values()) {
      if (hand.length !== 0) {
        return true;
      }
    }
    return false;
  }

  _tally_score_in_round(player_id) {
    var score_in_round = 0;
    for (const card of this.table.cards_taken.get(player_id)) {
      score_in_round += this.table.card_values.get(card);
    }
    return score_in_round;
  }

  _check_moonshot() {
    for (const player_id of this.player_ids) {
      var score_in_round = this._tally_score_in_round(player_id);
      if (score_in_round === 26) {
        return player_id;
      }
    }
    return null;
  }

  _finish_round() {
    const player_moonshot = this._check_moonshot();
    if (player_moonshot !== null) {
      this.round.player_moonshot = player_moonshot;
      console.log(
        `Player ${player_moonshot} shot the moon! Adding 26 to everyone else's score.`
      );

      for (const player_id of this.player_ids.filter(
        (p) => p !== player_moonshot
      )) {
        this.player_scores.set(
          player_id,
          this.player_scores.get(player_id) + 26
        );
      }
    } else {
      for (const [player_id, cards_taken] of this.table.cards_taken.entries()) {
        var score_in_round = this._tally_score_in_round(player_id);
        this.player_scores.set(
          player_id,
          this.player_scores.get(player_id) + score_in_round
        );
      }
    }
    console.log(
      `Round ${this.round_no} over, scores: ${[...this.player_scores]}`
    );
  }

  _game_continues() {
    const hundred_reached =
      Array.from(this.player_scores.values().filter((s) => s >= 100)).length !==
      0;
    const lowest_score = Math.min(...this.player_scores.values());
    const several_winners =
      Array.from(this.player_scores.values().filter((x) => x === lowest_score))
        .length > 1;
    if (hundred_reached & !several_winners) {
      console.log(`Game over! Scores: ${[...this.player_scores]}`);

      draw_status(`Game over!`);
      return false;
    }
    return true;
  }

  async play_round() {
    var round_scores = new Map();
    this.player_ids.forEach((player_id) => round_scores.set(player_id, 0));

    draw_nametags(this.player_scores, round_scores, this.player_ids_to_sides);

    this._start_round();

    this._draw_hands();

    const passed_cards = new Map();
    for (const [player_id, player] of this.players.entries()) {
      var pass_recipient = this.round.pass_direction.get(player_id);

      passed_cards.set(
        player_id,
        await player.pass_three_cards(
          this.table.hands.get(player_id),
          pass_recipient
        )
      );
      console.log(
        `${player_id} passed ${passed_cards.get(player_id)} to ${pass_recipient}`
      );
    }
    this._pass_cards(passed_cards);

    this._draw_hands();

    while (this._round_continues()) {
      this.player_ids.forEach((player_id) => {
        round_scores.set(player_id, this._tally_score_in_round(player_id));
      });
      draw_nametags(this.player_scores, round_scores, this.player_ids_to_sides);

      var player_idx = this._calculate_whose_move();
      // var player_idx = this.player_ids.indexOf(player_id)
      var player_order = [];
      var player_order_ids = [];
      for (var i = 0; i < 4; i++) {
        player_order.push((player_idx + i) % 4);
        player_order_ids.push(this.player_ids[(player_idx + i) % 4]);
      }

      console.log(
        `Starting trick ${this.round.trick_no}. Order: ${player_order_ids}`
      );

      for (var t = 0; t < 4; t++) {
        var player_idx = this._calculate_whose_move();
        var player_id = this.player_ids[player_idx];
        var player = this.players.get(player_id);
        console.log(`${player_id} moves...`);
        var played_card = await player.select_card_for_trick(
          this.table.hands.get(player_id),
          this._filter_illegal_options(
            this.table.hands.get(player_id),
            player_id
          ),
          this.round.current_trick,
          this.table.cards_taken
        );

        this._make_play(player_id, played_card);

        this._draw_hands();
        draw_trick_card(this.player_ids_to_sides[player_id], played_card);

        await new Promise((r) => setTimeout(r, 1000));
      }
      this._finish_trick();

      clear_table("trick");
    }
    this._finish_round();
  }

  async _draw_hands() {
    clear_table("hand");

    for (var [player_id, side] of Object.entries(this.player_ids_to_sides)) {
      draw_hand(
        this.table.hands.get(player_id),
        side,
        side === "bottom" ? false : true
      );
    }
  }

  async play() {
    while (this._game_continues()) {
      await this.play_round();
    }
  }
}
